unboundid-provider/src/main/java/org/ldaptive/provider/unboundid/UnboundIDProvider.java

```diff
@@ -83,6 +110,43 @@ protected LDAPConnectionOptions getDefaultLDAPConnectionOptions(final Connection
     if (cc.getResponseTimeout() != null) {
       options.setResponseTimeoutMillis(cc.getResponseTimeout().toMillis());
     }
+    // UnboundID does not do hostname verification by default
+    // set a default hostname verifier if no SSL socket verifier configured
+    final HostnameVerifier verifier;
+    if (cc.getSslConfig() != null && !cc.getSslConfig().isEmpty()) {
+      if (cc.getSslConfig().getHostnameVerifier() != null) {
+        verifier = new HostnameVerifierAdapter(cc.getSslConfig().getHostnameVerifier());
+      } else {
+        verifier = new DefaultHostnameVerifier();
+      }
+    } else {
+      verifier = new DefaultHostnameVerifier();
+    }
+    options.setSSLSocketVerifier(new SSLSocketVerifier()
+    {
+      /** Logger for this class. */
+      private  final Logger logger = LoggerFactory.getLogger(getClass());
+
+
+      @Override
+      public void verifySSLSocket(
+        final String host,
+        final int port,
+        final SSLSocket sslSocket)
+        throws LDAPException
+      {
+        logger.trace("Verifying SSLSocket {} for host {} with verifier {}", sslSocket, host, verifier);
+        try {
+          sslSocket.getSession().getPeerCertificates();
```

Comment:
For your consideration:

        logger.trace("Verifying SSLSocket {} for host {} with verifier {}", sslSocket, host, verifier);
        final SSLSession session = sslSocket.getSession();
        try {  
          // Following call is required to initiate SSL handshake
          session.getPeerCertificates();
          if (!verifier.verify(host, session)) {
            throw new LDAPException(ResultCode.CONNECT_ERROR, "Hostname verification failed for " + host);
          }
        } catch (SSLPeerUnverifiedException e) {
          logger.trace("Skipped hostname verification for SSLSocket {} due to handshake failure", sslSocket, e);
        }

Reply:
Sure. I think a better question is whether hostname verification should succeed when trust fails. I found the errors confusing in that case, which led to this implementation. In hindsight, it may be clearer to simply throw in that case.

Reply:
Agree that throwing is better on the principle that you really ought to have an impeccable reason for swallowing exceptions. Your ambivalence about the behavior suggests you do not, so throwing is probably the safest bet.